// @aave/contract-helpers/dist/cjs/umbrella/RewardsDistributor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RewardsDistributorService = void 0;
const ethers_1 = require("ethers");
const types_1 = require("../commons/types");
const utils_1 = require("../commons/utils");
const IRewardsDistributor__factory_1 = require("./typechain/IRewardsDistributor__factory");
class RewardsDistributorService {
    constructor(rewardsDistributorAddress) {
        this.rewardsDistributorAddress = rewardsDistributorAddress;
        this.interface = IRewardsDistributor__factory_1.IRewardsDistributor__factory.createInterface();
    }
    // Claim all rewards across all stake tokens
    claimAllAvailableRewards({ stakeTokens, sender, }) {
        const tx = {};
        const receiver = sender;
        tx.data = this.interface.encodeFunctionData('claimAllRewards(address[],address)', [stakeTokens, receiver]);
        tx.from = sender;
        tx.to = this.rewardsDistributorAddress;
        tx.gasLimit = ethers_1.BigNumber.from(utils_1.gasLimitRecommendations[types_1.ProtocolAction.umbrellaClaimAllRewards]
            .recommended);
        return tx;
    }
    // Claim all rewards for a specific stake token
    claimAllRewards({ stakeToken, sender, }) {
        const tx = {};
        const receiver = sender;
        const txData = this.interface.encodeFunctionData('claimAllRewards(address,address)', [stakeToken, receiver]);
        tx.data = txData;
        tx.from = sender;
        tx.to = this.rewardsDistributorAddress;
        tx.gasLimit = ethers_1.BigNumber.from(utils_1.gasLimitRecommendations[types_1.ProtocolAction.umbrellaClaimAllRewards]
            .recommended);
        return tx;
    }
    claimSelectedRewards({ stakeToken, rewards, sender, }) {
        const tx = {};
        const receiver = sender;
        const txData = this.interface.encodeFunctionData('claimSelectedRewards(address,address[],address)', [stakeToken, rewards, receiver]);
        tx.data = txData;
        tx.from = sender;
        tx.to = this.rewardsDistributorAddress;
        tx.gasLimit = ethers_1.BigNumber.from(utils_1.gasLimitRecommendations[types_1.ProtocolAction.umbrellaClaimSelectedRewards]
            .recommended);
        return tx;
    }
}
exports.RewardsDistributorService = RewardsDistributorService;
