// @aave/contract-helpers/dist/cjs/umbrella/StakeToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StakeTokenService = void 0;
const bytes_1 = require("@ethersproject/bytes");
const ethers_1 = require("ethers");
const types_1 = require("../commons/types");
const utils_1 = require("../commons/utils");
const StakeToken__factory_1 = require("./typechain/StakeToken__factory");
class StakeTokenService {
    constructor(stakeTokenAddress) {
        this.stakeTokenAddress = stakeTokenAddress;
        this.interface = StakeToken__factory_1.IERC4626StakeToken__factory.createInterface();
    }
    cooldown({ sender }) {
        const tx = {};
        const txData = this.interface.encodeFunctionData('cooldown');
        tx.data = txData;
        tx.from = sender;
        tx.to = this.stakeTokenAddress;
        tx.gasLimit = ethers_1.BigNumber.from(utils_1.gasLimitRecommendations[types_1.ProtocolAction.umbrellaStakeTokenCooldown]
            .recommended);
        return tx;
    }
    deposit({ amount, sender }) {
        const tx = {};
        const receiver = sender;
        const txData = this.interface.encodeFunctionData('deposit', [
            amount,
            receiver,
        ]);
        tx.data = txData;
        tx.from = sender;
        tx.to = this.stakeTokenAddress;
        tx.gasLimit = ethers_1.BigNumber.from(utils_1.gasLimitRecommendations[types_1.ProtocolAction.umbrellaStakeTokenDeposit]
            .recommended);
        return tx;
    }
    depositWithPermit({ amount, deadline, permit, sender, }) {
        const tx = {};
        const signature = (0, bytes_1.splitSignature)(permit);
        const receiver = sender;
        const txData = this.interface.encodeFunctionData('depositWithPermit', [
            amount,
            receiver,
            deadline,
            { v: signature.v, r: signature.r, s: signature.s },
        ]);
        tx.data = txData;
        tx.from = sender;
        tx.to = this.stakeTokenAddress;
        tx.gasLimit = ethers_1.BigNumber.from(utils_1.gasLimitRecommendations[types_1.ProtocolAction.umbrellaStakeTokenDepositWithPermit].recommended);
        return tx;
    }
    redeem({ amount, sender }) {
        const tx = {};
        const receiver = sender;
        const owner = sender;
        const txData = this.interface.encodeFunctionData('redeem', [
            amount,
            receiver,
            owner,
        ]);
        tx.data = txData;
        tx.from = sender;
        tx.to = this.stakeTokenAddress;
        tx.gasLimit = ethers_1.BigNumber.from(utils_1.gasLimitRecommendations[types_1.ProtocolAction.umbrellaStakeTokenRedeem]
            .recommended);
        return tx;
    }
}
exports.StakeTokenService = StakeTokenService;
