// @aave/contract-helpers/dist/cjs/umbrella/UmbrellaBatchHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UmbrellaBatchHelperService = void 0;
const bytes_1 = require("@ethersproject/bytes");
const ethers_1 = require("ethers");
const types_1 = require("../commons/types");
const utils_1 = require("../commons/utils");
const UmbrellaBatchHelper__factory_1 = require("./typechain/UmbrellaBatchHelper__factory");
class UmbrellaBatchHelperService {
    constructor(umbrellaBatchHelperAddress) {
        this.umbrellaBatchHelperAddress = umbrellaBatchHelperAddress;
        this.interface = UmbrellaBatchHelper__factory_1.UmbrellaBatchHelper__factory.createInterface();
    }
    deposit({ sender, stakeToken, amount, edgeToken, }) {
        const txData = this.interface.encodeFunctionData('deposit', [
            {
                edgeToken,
                stakeToken,
                value: amount,
            },
        ]);
        const data = txData;
        const from = sender;
        const to = this.umbrellaBatchHelperAddress;
        const gasLimit = ethers_1.BigNumber.from(utils_1.gasLimitRecommendations[types_1.ProtocolAction.umbrellaStakeGatewayStake]
            .recommended);
        const tx = {
            data,
            from,
            to,
            gasLimit,
        };
        return tx;
    }
    depositWithPermit({ sender, edgeToken, stakeToken, amount, deadline, permit, }) {
        const signature = (0, bytes_1.splitSignature)(permit);
        const permitTxData = this.interface.encodeFunctionData('permit', [
            {
                token: edgeToken,
                value: amount,
                deadline,
                v: signature.v,
                r: signature.r,
                s: signature.s,
            },
        ]);
        const { data: depositTxData } = this.deposit({
            sender,
            stakeToken,
            amount,
            edgeToken,
        });
        const txData = this.interface.encodeFunctionData('multicall', [
            [permitTxData, depositTxData],
        ]);
        const data = txData;
        const from = sender;
        const to = this.umbrellaBatchHelperAddress;
        const gasLimit = ethers_1.BigNumber.from(utils_1.gasLimitRecommendations[types_1.ProtocolAction.umbrellaStakeGatewayStake]
            .recommended);
        const tx = {
            data,
            from,
            to,
            gasLimit,
        };
        return tx;
    }
    redeem({ sender, stakeToken, amount, edgeToken, }) {
        const txData = this.interface.encodeFunctionData('redeem', [
            {
                edgeToken,
                stakeToken,
                value: amount,
            },
        ]);
        const data = txData;
        const from = sender;
        const to = this.umbrellaBatchHelperAddress;
        const gasLimit = ethers_1.BigNumber.from(utils_1.gasLimitRecommendations[types_1.ProtocolAction.umbrellaStakeGatewayRedeem]
            .recommended);
        const tx = {
            data,
            from,
            to,
            gasLimit,
        };
        return tx;
    }
    redeemWithPermit({ sender, edgeToken, stakeToken, amount, deadline, permit, }) {
        const signature = (0, bytes_1.splitSignature)(permit);
        const permitTxData = this.interface.encodeFunctionData('permit', [
            {
                token: edgeToken,
                value: amount,
                deadline,
                v: signature.v,
                r: signature.r,
                s: signature.s,
            },
        ]);
        const { data: redeemTxData } = this.redeem({
            sender,
            stakeToken,
            amount,
            edgeToken,
        });
        const txData = this.interface.encodeFunctionData('multicall', [
            [permitTxData, redeemTxData],
        ]);
        const data = txData;
        const from = sender;
        const to = this.umbrellaBatchHelperAddress;
        const gasLimit = ethers_1.BigNumber.from(utils_1.gasLimitRecommendations[types_1.ProtocolAction.umbrellaStakeGatewayStake]
            .recommended);
        const tx = {
            data,
            from,
            to,
            gasLimit,
        };
        return tx;
    }
}
exports.UmbrellaBatchHelperService = UmbrellaBatchHelperService;
